### ISSUE TITLE
Run Kmonad on startup

### ISSUE THREAD

**Issue by austinwilcox** [kmonad/kmonad#482]
Hi,
Is it possible to set up kmonad as a cron job to run at system startup? I have tried editing the root cron jobs with 
```sudo crontab -e``` 
and I have the following line as a job to run 
```@reboot kmonad /home/<username>/.dotfiles/kmonad-layouts/colemak-dh-extend-ansi.kbd &```
but I get nothing so far. Is there an easier way to do this? I use colemak dh on my laptop and I would prefer to have colemak work on system startup instead of booting up a scratchpad each time and running kmonad.
Note: This is on Linux

**Comment by salastro** [comment]
Is not it possible to just add it to `.xinitrc` or as an [init system service](https://github.com/kmonad/kmonad/tree/master/startup) to start on boot?

**Comment by MuhammedZakir** [comment]
> Have you looked through the documentation or the issues? IIRC there have been plenty of posts about running kmonad as an init service (for various init systems), as a cron job, on login, etc

Also, when something goes wrong, adding `--log debug` flag to kmonad command really helps. Almost all of the time. you can infer what could be the reason from the output.

**Comment by mnesarco** [comment]
On Linux you can launch KMonad when your keyboard is plugged in (also detected as startup) using udev:

## 1. Install at
**at** is a linux command to schedule jobs.

## 2. Create a launcher (ie /opt/start_km.sh)
```
#!/bin/bash

# Change <username> to your real user and any path to your real paths

NH=/usr/bin/nohup
KEYBOARD=/home/<username>/.dotfiles/kmonad-layouts/my-custom-layout.kbd
KMONAD="/usr/bin/kmonad"

echo "sleep 1 ; $NH $KMONAD $KEYBOARD >/dev/null 2>&1 &" | at now

```

Be sure to make the file executable ( `chmod +x /opt/start_km.sh` )


## 3. Create some udev rules (/etc/udev/rules.d)

```
# file: /etc/udev/rules.d/99-my-kmonad.rules

# Change attributes to match your keyboard hardware. This is an example from my custom keyboard:
# Change <username> to your real user

# Create a well known dev name for your keyboard
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="Marz Keyboard LR1", ATTRS{phys}=="marz-kbd-lr", SYMLINK+="marz-kbd-lr"

# Launch kmonad
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="Marz Keyboard LR1", ATTRS{phys}=="marz-kbd-lr", ENV{DISPLAY}=":0", ENV{XAUTHORITY}="/home/<username>/.Xauthority", ENV{HOME}="/home/<username>", RUN+="/opt/start_km.sh"

```

Copy your rules file into `/etc/udev/rules.d/`

## 4 Reload udev

```
udevadm control --reload
```

## 5 Unplug/Plug your keyboard

Everytime you plug your keyboard it will be detected by udev and will start kmonad. Also works on boot.

**Comment by austinwilcox** [comment]
> Is not it possible to just add it to `.xinitrc` or as an [init system service](https://github.com/kmonad/kmonad/tree/master/startup) to start on boot?

So I have attempted this and I'm not having any luck getting the service started that way.

**Comment by mnesarco** [comment]
@austinwilcox I am using this setup with my DIY cheap split keyboard + udev + kmonad:


![WhatsApp Image 2022-02-06 at 12 51 14 PM](https://user-images.githubusercontent.com/1087317/154128317-1dbeef6a-44b4-4f60-929b-8587a4b2cabf.jpeg)

this is my udev rules file:

 ```
# Rules for Marz Split Keyboard

# Left and right source keyboards
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="SEMICO USB Keyboard", ATTRS{phys}=="usb-0000:00:14.0-2.4/input0", SYMLINK+="marz-kbd-l"
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="SEMICO USB Keyboard", ATTRS{phys}=="usb-0000:00:14.0-2.3/input0", SYMLINK+="marz-kbd-r"
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="SEMICO USB Keyboard", ATTRS{phys}=="usb-0000:00:14.0-2.[34]/input0", RUN+="/opt/marz/kbd/marz_kbd.sh"

# Joined LR target keyboard
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="Marz Keyboard LR1", ATTRS{phys}=="marz-kbd-lr", SYMLINK+="marz-kbd-lr"
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="Marz Keyboard LR1", ATTRS{phys}=="marz-kbd-lr", ENV{DISPLAY}=":0", ENV{XAUTHORITY}="/home/xxx/.Xauthority", ENV{HOME}="/home/xxx", RUN+="/opt/marz/kbd/marz_kmonad.sh"

# KMonad keyboard
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="marz-kbd-kmonad", SYMLINK+="marz-kbd-kmonad"
SUBSYSTEM=="input", ACTION=="add", ATTRS{name}=="marz-kbd-kmonad", ENV{DISPLAY}=":0", ENV{XAUTHORITY}="/home/xxx/.Xauthority", ENV{HOME}="/home/xxx", RUN+="/opt/marz/kbd/marz_setxkbmap.sh"
```

This is a very specific scenario, but you get the idea.

In my case when I plug my two one-handed keyboards in specific ports, they are recognized and joined into a single virtual keyboard used by kmonad. Then kmomad is launched and the rest is magic.

**Comment by austinwilcox** [comment]
I have an update on this now, I did figure out how to get it to run at startup.

@mnesarco This guide you provided is incredible! I am going to try and follow these steps to get my external keyboard to connect and auto use colemak.

I would have ended up using your solution posted above, but I did some more digging into my system and I realized that I didn't have a process that would run my cron jobs. So I installed cronie and modified my script to look like this:
```
@reboot sudo /usr/bin/kmonad /home/<username>/.dotfiles/kmonad-layouts/colemak-dh-extend-ansi.kbd
```
As soon as my computer gets to the login screen, my laptop keyboard is using colemak keybindings.

Thanks for all the fantastic community support!
